gui/elements/qt/_layerList.py

```diff
@@ -0,0 +1,32 @@
+from PyQt5.QtCore import Qt
+from PyQt5.QtWidgets import QWidget, QVBoxLayout, QScrollArea
+
+
+class QtLayerList(QScrollArea):
+    def __init__(self):
+        super().__init__()
+
+        self.setWidgetResizable(True)
+        #self.setFixedWidth(315)
+        self.setHorizontalScrollBarPolicy(Qt.ScrollBarAlwaysOff)
+        scrollWidget = QWidget()
+        self.setWidget(scrollWidget)
+        self.layersLayout = QVBoxLayout(scrollWidget)
+        self.layersLayout.addStretch(1)
+
+    def add(self, index, layer):
+        if layer._qt is not None:
+            self.layersLayout.insertWidget(index, layer._qt)
+
+    def remove(self, layer):
+        if layer._qt is not None:
+            self.layersLayout.removeWidget(layer._qt)
+            layer._qt.deleteLater()
+            layer._qt = None
+
+    def reorder(self, layerList):
+        for i in range(len(layerList)):
+            layer = layerList[i]
+            if layer._qt is not None:
+                self.layersLayout.removeWidget(layer._qt)
+                self.layersLayout.insertWidget(i,layer._qt)
```

Comment:
Hmm, it's not clear to me why this operation would fail with an ordinary list. The for loop ensures that I visit all the items in the list (assuming they are unique) and the combination of the remove and insert always ensures that I move each item to it's new location in the list.

I've checked it a bunch of times doing pairwise swaps using the `swap` function and it works fine. I just tried doing a `reorder` but I got an error
```python
NameError                                 Traceback (most recent call last)
<ipython-input-19-8cc2f5393ffa> in <module>()
----> 1 viewer.layers.reorder([2,1,4,3])

~/Github/napari-gui/napari_gui/elements/_layer_list.py in reorder(self, *ordering)
    252         if not isinstance(ordering[0], (int, Layer)):
    253             ordering = ordering[0]
--> 254             if not isinstance(ordering, Sequence):
    255                 raise TypeError(f'expected {ordering} to be Sequence; '
    256                                 f'got {type(ordering)}') from None

NameError: name 'Sequence' is not defined
```
not sure where that import needs to come from

Reply:
`from collections import Sequence`

I agree with @kne42 that this code is difficult to parse. Would you mind adding some documentation (either a docstring or comments) about what `reorder` should achieve? ie what is the input and what is the state of the output?

Specifically, does the input need to include all the existing layers, or only the ones being reordered? If the former, then I would suggest just clearing the view and then appending them in sequence. If the latter, then it feels weird to me that they will move to the front (unless I'm misreading, which is possible).

Reply:
Yeah, after looking at it again I understand why it works, but it's not super obvious.

Also it should be <code>from collections<strong>.abc</strong> import Sequence</code> (Python 3.7 introduces a deprecation warning for this).

Reply:
I also agree that it is not super clear but we can refactor later it, i believe.

Reply:
> we can refactor later

This is only true if everyone involved understands what it's trying to do. =)

> should be from collections.abc import Sequence

Ugh, as if it wasn't unclear enough. LOL But I get why they changed it. Thanks for the pointer, @kne42!

Reply:
@jni the input contains all the existing layers and the list is entirely rebuilt in the for loop. I guess what you're advocating for would be to put the clear / remove command outside the for loop and then rebuild from an empty slate.